source/common/network/dns_resolver/dns_factory.h

```diff
@@ -0,0 +1,122 @@
+// 3.3) if config is DnsFilterConfig, pack an empty CaresDnsResolverConfig into
+//      typed_dns_resolver_config.
+// 3.4) For all others, copy config.use_tcp_for_dns_lookups into
+//      CaresDnsResolverConfig.dns_resolver_options.use_tcp_for_dns_lookups
+// 3.5) For ClusterConfig, one extra thing is to copy dns_resolvers into
+//      CaresDnsResolverConfig.resolvers,
+// 3.6) Then pack CaresDnsResolverConfig into typed_dns_resolver_config.
+
+template <class T>
+void makeDnsResolverConfig(
+    const T& config, envoy::config::core::v3::TypedExtensionConfig& typed_dns_resolver_config) {
+  // If the config has typed_dns_resolver_config, and the corresponding DNS resolver factory
+  // is registered, copy it into typed_dns_resolver_config and return.
+  if (config.has_typed_dns_resolver_config()) {
+    Network::DnsResolverFactory* dns_resolver_factory;
+    dns_resolver_factory = Config::Utility::getAndCheckFactory<Network::DnsResolverFactory>(
+        config.typed_dns_resolver_config(), true);
+    if ((dns_resolver_factory != nullptr) &&
+        (dns_resolver_factory->category() == dns_resolver_category)) {
+      typed_dns_resolver_config.MergeFrom(config.typed_dns_resolver_config());
+      return;
+    }
+  }
+
+  // If Envoy is built in MacOS and "envoy.dns_resolver.apple" extension is enabled in build file,
+  // and  the run time flag envoy.restart_features.use_apple_api_for_dns_lookups is enabled,
+  // in this case, use Apple DNS library. To achieve this, Crafting a AppleDnsResolverConfig
+  // object and pack into typed_dns_resolver_config.
+  if (Config::Utility::getAndCheckFactoryByName<Network::DnsResolverFactory>(apple_dns_resolver,
+                                                                             true)) {
+    static bool use_apple_api_for_dns_lookups =
+        Runtime::runtimeFeatureEnabled("envoy.restart_features.use_apple_api_for_dns_lookups");
+    if (use_apple_api_for_dns_lookups) {
+      envoy::extensions::network::dns_resolver::apple::v3::AppleDnsResolverConfig apple;
+      // Pack AppleDnsResolverConfig object into typed_dns_resolver_config.
+      typed_dns_resolver_config.mutable_typed_config()->PackFrom(apple);
+      typed_dns_resolver_config.set_name(apple_dns_resolver);
+      return;
+    }
+  }
+  // Otherwise, fall back to default behavior.
+  envoy::extensions::network::dns_resolver::cares::v3::CaresDnsResolverConfig cares;
+  if (config.has_dns_resolution_config()) {
+    // Copy resolvers if config has it.
+    if (!config.dns_resolution_config().resolvers().empty()) {
+      cares.mutable_resolvers()->MergeFrom(config.dns_resolution_config().resolvers());
+    }
+    cares.mutable_dns_resolver_options()->MergeFrom(
+        config.dns_resolution_config().dns_resolver_options());
+  } else {
+    // Skipping copying these fields for DnsFilterConfig.
+    if constexpr (!(std::is_same_v<T, envoy::extensions::filters::udp::dns_filter::v3alpha::
```

Comment:
this pattern makes me wonder if we should use some other OO features rather than templatization to achieve this polymorphism, but I admit a better pattern doesn't come to mind immediately. But it reminds me of having to do dynamic_cast to conditionalize an implementation in a virtual method.

Reply:
I agree that this function became difficult to read and it would be good to refactor out some common parts, like apple related logic and when there is typed_dns_resolver_config. It may good to move helper function into the .cc file as well.

Alternative approach to using template and then `if constexpr` is to provide overloads for each proto that needs to deal with the backward compatibility. This maybe closer to home for C++

Reply:
I went through the overloads approach, but the bootstrap.proto and dns_cache.proto have identical logic, which lead to identical code. So I would like to keep using template + template specification to implement the special behavior for dns_filter.proto and cluster.proto. This way there is no such std::is_same_v<> check. 
Please take a look the latest code and let me know whether it looks good to you.